// rql/js-array.js
/*
 * An implementation of RQL for JavaScript arrays. For example:
 * require("./js-array").query("a=3", {}, [{a:1},{a:3}]) -> [{a:3}]
 *
 */



({define:typeof define!="undefined"?define:function(deps, factory){module.exports = factory(exports, require("./parser"), require("./query"));}}).
define(["exports", "./parser", "./query"], function(exports, parser, QUERY){
//({define:typeof define!="undefined"?define:function(deps, factory){module.exports = factory(exports, require("./parser"));}}).
//define(["exports", "./parser"], function(exports, parser){

var parseQuery = parser.parseQuery;
var stringify = JSON.stringify || function(str){
	return '"' + str.replace(/"/g, "\\\"") + '"';
};
exports.jsOperatorMap = {
	"eq" : "===",
	"ne" : "!==",
	"le" : "<=",
	"ge" : ">=",
	"lt" : "<",
	"gt" : ">"
};

function isPresent(path, items){
	var res = [];
	//console.log("is present : "+name)
	items.forEach(function(e){	
	//	console.log("got items : test : " + name + " - object "+JSON.stringify(e));
		if(retrieve(e, path))
			res.push(e);
	});
	return res;
}


function retrieve(object, path)
{
	var parts = [];
	//console.log("retrieve path : " +path + " - "+object)
	if(path)
		if(typeof path  === 'string')
			parts = path.split(".");
		else if(path instanceof Array)
			parts = path;
	if(objectPrefix && objectPrefix != "")
		parts.unshift(objectPrefix);
	//console.log("retrieve final path : " +JSON.stringify(parts))

	var tmp = object;
	while(parts.length>1)
	{
		var part = parts.shift();
		if(!tmp[part])
			return null;
		tmp = tmp[part];
	}
	var res = null;

	//console.log("retrieve last path got : "+JSON.stringify(parts))
	if(tmp)
		res = tmp[parts.shift()];
	///console.log("*****retrieve path got : "+res)
	return res;
}

exports.operators = {
	sort: function(){
		var terms = [];
		for(var i = 0; i < arguments.length; i++){
			var sortAttribute = arguments[i];
			var firstChar = sortAttribute.charAt(0);
			var term = {attribute: sortAttribute, ascending: true};
			if (firstChar == "-" || firstChar == "+") {
				if(firstChar == "-"){
					term.ascending = false;
				}
				term.attribute = term.attribute.substring(1);
			}
			terms.push(term);
		}
		this.sort(function(a, b){
			for (var term, i = 0; term = terms[i]; i++) {
				var ar = retrieve(a, term.attribute);
				var br = retrieve(b, term.attribute);
				if (ar != br) {
					return term.ascending == ar > br ? 1 : -1;
				}
			}
			return 0;
		});
		return this;
	},
	match: filter(function(value, regex){
		return new RegExp(regex).test(retrieve(value));
	}),
	"in": filter(function(value, values){
		//console.log("IN OP : query "+value+" in : "+JSON.stringify(values))
		return values.indexOf(value) > -1;
	}),
	out: filter(function(value, values){
		return values.indexOf(retrieve(value)) == -1;
	}),
	contains: filter(function(array, value){
		if(typeof value == "function"){
			return array instanceof Array && array.some(function(v){
				var r = retrieve(value);
				if(r)
					return r.call([v]).length;
				return false;
			});
		}
		else{
			return array instanceof Array && array.indexOf(retrieve(value)) > -1;
		}
	}),
	excludes: filter(function(array, value){
		if(typeof value == "function"){
			return !array.some(function(v){
				var r = retrieve(value);
				if(r)
					return r.call([v]).length;
				return false;
			});
		}
		else{
			return array.indexOf(value) == -1;
		}
	}),
	or: function(){
		var items = [];
		//TODO: remove duplicates and use condition property
		for(var i = 0; i < arguments.length; i++){
			if(typeof arguments[i] == 'function')
				items = items.concat(arguments[i].call(this));
			else
				items = items.concat(isPresent(arguments[i], items))

		}
		return items;
	},
	and: function(){
		var items = this;
		// TODO: use condition property
		for(var i = 0; i < arguments.length; i++){
			///console.log("and call args : "+i+ " - "+arguments[i]);
			if(typeof arguments[i] == 'function')
				items = arguments[i].call(items);
			else //if(typeof arguments[i] === 'string')
				items = isPresent(arguments[i], items);
		}
		return items;
	},
	select: function(){
		var args = arguments;
		var argc = arguments.length;
		return this.map(function(object){
			var selected = {};
			for(var i = 0; i < argc; i++){
				var propertyName = args[i];
				var value = evaluateProperty(object, propertyName);
				if(typeof value != "undefined"){
					selected[propertyName] = value;
				}
			}
			return selected;
		});
	},
	/* _______________________________________________ WARNING : NOT IMPLEMENTED WITH PREFIX*/
	unselect: function(){
		var args = arguments;
		var argc = arguments.length;
		return this.map(function(object){
			var selected = {};
			for (var i in object)
			 if (object.hasOwnProperty(i)) {
				selected[i] = object[i];
			}
			for(var i = 0; i < argc; i++) {
				delete selected[args[i]];
			}
			return selected;
		});
	},

	values: function(first){
		if(arguments.length == 1){
			return this.map(function(object){
				return retrieve(object, first) ;
			});
		}
		var args = arguments;
		var argc = arguments.length;
		return this.map(function(object){
			var realObject = retrieve(object);
			var selected = [];
			if (argc === 0) {
				for(var i in realObject) if (realObject.hasOwnProperty(i)) {
					selected.push(realObject[i]);
				}
			} else {
				for(var i = 0; i < argc; i++){
					var propertyName = args[i];
					selected.push(realObject[propertyName]);
				}
			}
			return selected;
		});
	},
	limit: function(limit, start, maxCount){
		var totalCount = this.length;
		start = start || 0;
		var sliced = this.slice(start, start + limit);
		if(maxCount){
			sliced.start = start;
			sliced.end = start + sliced.length - 1;
			sliced.totalCount = Math.min(totalCount, typeof maxCount === "number" ? maxCount : Infinity);
		}
		return sliced;
	},
	distinct: function(){

		var primitives = {};
		var needCleaning = [];
		var newResults = this.filter(function(value){
			value = retrieve(value);
			//console.log("distinct : value : "+value);
			if(value && typeof value == "object"){
				if(!value.__found__){
					value.__found__ = function(){};// get ignored by JSON serialization
					needCleaning.push(value);
					return true;
				}
				return false;
			}else{
				//console.log("found primitive : "+value + " - "+JSON.stringify(primitives))
				if(!primitives[value]){
					//console.log("primitive was not there")
					primitives[value] = true;
					return true;
				}
				return false;
			}
		});
		needCleaning.forEach(function(object){
			delete object.__found__;
		});
		return newResults;
	},
	recurse: function(property){
		// TODO: this needs to use lazy-array
		var newResults = [];
		function recurse(value){
			if(value instanceof Array){
				value.forEach(recurse);
			}else{
				newResults.push(value);
				if(property){
					value = value[property];
					if(value && typeof value == "object"){
						recurse(value);
					}
				}else{
					for(var i in value){
						if(value[i] && typeof value[i] == "object"){
							recurse(value[i]);
						}
					}
				}
			}
		}
		recurse(retrieve(this));
		return newResults;
	},
	aggregate: function(){
		var distinctives = [];
		var aggregates = [];
		for(var i = 0; i < arguments.length; i++){
			var arg = arguments[i];
			if(typeof arg === "function"){
				 aggregates.push(arg);
			}else{
				distinctives.push(arg);
			}
		}
		var distinctObjects = {};
		var dl = distinctives.length;
		this.forEach(function(object){
			object = retrieve(object);
			var key = "";
			for(var i = 0; i < dl;i++){
				key += '/' + object[distinctives[i]];
			}
			var arrayForKey = distinctObjects[key];
			if(!arrayForKey){
				arrayForKey = distinctObjects[key] = [];
			}
			arrayForKey.push(object);
		});
		var al = aggregates.length;
		var newResults = [];
		for(var key in distinctObjects){
			var arrayForKey = distinctObjects[key];
			var newObject = {};
			for(var i = 0; i < dl;i++){
				var property = distinctives[i];
				newObject[property] = arrayForKey[0][property];
			}
			for(var i = 0; i < al;i++){
				var aggregate = aggregates[i];
				newObject[i] = aggregate.call(arrayForKey);
			}
			newResults.push(newObject);
		}
		return newResults;
	},
	between: filter(function(value, range){
		value = retrieve(value);
		return value >= range[0] && value < range[1];
	}),
	sum: reducer(function(a, b){
		var a= retrieve(a);
		var b = retrieve(b);
		return a + b;
	}),
	mean: function(property){
		return exports.operators.sum.call(this, property)/this.length;
	},
	max: reducer(function(a, b){
		var a= retrieve(a);
		var b = retrieve(b);
		return Math.max(a, b);
	}),
	min: reducer(function(a, b){
		var a= retrieve(a);
		var b = retrieve(b);
		return Math.min(a, b);
	}),
	count: function(){

		return this.length;
	},
	first: function(){
		return this[0];
	},
	one: function(){
		if(this.length > 1){
			throw new TypeError("More than one object found");
		}
		return this[0];
	}
};
exports.filter = filter;
function filter(condition, not){
	// convert to boolean right now
	var filter = function(property, second){
		if(typeof second == "undefined"){
			second = property;
			property = undefined;
		}
		var args = arguments;
		var filtered = [];
		for(var i = 0, length = this.length; i < length; i++){
			var item = this[i];
			if(condition(evaluateProperty(item, property), second)){
				filtered.push(item);
			}
		}
		return filtered;
	};
	filter.condition = condition;
	return filter;
};
function reducer(func){
	return function(property){
		if(property){
			return this.map(function(object){
				return retrieve(object, property);
			}).reduce(func);
		}else{
			return this.reduce(func);
		}
	}
}
exports.evaluateProperty = evaluateProperty;
function evaluateProperty(object, property){
	if(property instanceof Array){
		return retrieve(object, decodeURIComponent(property));
	}else if(typeof property == "undefined"){
		return retrieve(object);
	}else{
		return retrieve(object, decodeURIComponent(property));
	}
};
var conditionEvaluator = exports.conditionEvaluator = function(condition){
	var jsOperator = exports.jsOperatorMap[term.name];
	if(jsOperator){
		js += "(function(item){return item." + term[0] + jsOperator + "parameters[" + (index -1) + "][1];});";
	}
	else{
		js += "operators['" + term.name + "']";
	}
	return eval(js);
};
exports.executeQuery = function(query, options, target){
	return exports.query(query, options, target);
}
exports.query = query;
exports.missingOperator = function(operator){
	throw new Error("Operator " + operator + " is not defined");
}
var objectPrefix = null;
function query(query, options, target, prefix){
	options = options || {};
	query = parseQuery(query, options.parameters);
	//console.log("query : "+JSON.stringify(query) + " - prefix : "+prefix);
	//console.log("target : "+JSON.stringify(target));
	prefix = prefix || "";
	objectPrefix = prefix || null;
	function t(){}
	t.prototype = exports.operators;
	var operators = new t;
	// inherit from exports.operators
	for(var i in options.operators){
		operators[i] = options.operators[i];
	}
	function op(name){
		return operators[name]||exports.missingOperator(name);
	}
	var parameters = options.parameters || [];
	var js = "";
	function queryToJS(value){
		if(value && typeof value === "object"){
			if(value instanceof Array){
				//console.log("__________________ QUERY TO JS : query object is array : "+JSON.stringify(value))
				return '[' + value.map(queryToJS) + ']';
			}else{
				//console.log("QUERY TO JS : query value : "+JSON.stringify(value))
				var jsOperator = exports.jsOperatorMap[value.name];
				if(jsOperator)
				{
				//	console.log("__________________ GOT OPERTAOR : "+value.name + " - have prefix ? "+prefix)
					///console.log("item['foo.bar'] ==> (item && item.foo && item.foo.bar && ...)")
					// item['foo.bar'] ==> (item && item.foo && item.foo.bar && ...)
					var path = value.args[0];
					var target = value.args[1];
					if (typeof target == "undefined"){
				//		console.log("Query to js : path is not provided : "+JSON.stringify(path))
						var item = "item";
						target = path;
						if(objectPrefix && objectPrefix != "")
							item += "."+objectPrefix;
					}
					else {
				//		console.log("Query to js : path will be splitted : "+JSON.stringify(path))

						if(typeof path === 'string')
							path = path.split(".");
				//		console.log("Query to js : path have been splitted : "+JSON.stringify(path))

						var item = "item";
						var escaped = [];
						if(prefix && prefix != "")
							path.unshift(prefix);
				//		console.log("Query to js : path is array : "+JSON.stringify(path))
						for(var i = 0;i < path.length; i++){
							escaped.push(stringify(path[i]));
							item +="&&item[" + escaped.join("][") + ']';
						}
					}
					// use native Array.prototype.filter if available
					var condition = item + jsOperator + queryToJS(target);
				//	console.log("conditions : "+condition)
					if (typeof Array.prototype.filter === 'function') {
						return "(function(){return this.filter(function(item){return " + condition + "})})";
						//???return "this.filter(function(item){return " + condition + "})";
					} else {
						throw "error : array have no filter";
						return "(function(){var filtered = []; for(var i = 0, length = this.length; i < length; i++){var item = this[i];if(" + condition + "){filtered.push(item);}} return filtered;})";
					}
				}else{
				//	console.log("NO JS OPERATOR")
					return "(function(){return op('" + value.name + "').call(this" +
						(value && value.args && value.args.length > 0 ? (", " + value.args.map(queryToJS).join(",")) : "") +
						")})";
				}
			}
		}else{
			return typeof value === "string" ? stringify(value) : value;
		}
	}
	var evaluationString = "(function(target){return " + queryToJS(query) + ".call(target);})";
	var evaluator = eval(evaluationString);
	//console.log("EVALUATION STRING \n\n"+evaluationString)
	return target ? evaluator(target) : evaluator;
}
function throwMaxIterations(){
	throw new Error("Query has taken too much computation, and the user is not allowed to execute resource-intense queries. Increase maxIterations in your config file to allow longer running non-indexed queries to be processed.");
}
exports.maxIterations = 10000;
return exports;
});
